### DISCUSSION TITLE
We can log in to Bentley through api? is it possible?🙏

### DISCUSSION THREAD

**Discussion by jagadishjay** [iTwin/itwinjs-core#3901] [General]
Can we skip log in page in UI and log in through API is it possible?

**Comment by aruniverse** [comment]
The closest thing to what you're looking for is probably using a proxy token server. Here's an example:

https://github.com/calebmshafer/itwin-viewer-proxy

**Reply by jagadishjay** [reply]
While installing this https://github.com/calebmshafer/itwin-viewer-proxy
I'm getting errors can you have any doc for step by step.

**Reply by pmconne** [reply]
> I'm getting errors

Show the output.
I installed token-server and react-viewer. Aside from some npm audit warnings they installed cleanly. Node 16.15.1. npm 8.11.0.

**Reply by jagadishjay** [reply]
Can you have git repo? I'll refer that.

**Reply by pmconne** [reply]
Sorry, what?
https://github.com/calebmshafer/itwin-viewer-proxy is a git repo. It contains a README and two packages. Each of those packages have their own READMEs. Did you read them?
If you are having trouble, be specific about what you need help with, what you did, any errors you received, etc.

**Reply by rakesh-agathsya** [reply]
Hi @pmconne Thanks for the response.

I'll explain the situation. We have an application under development which uses Bentley to process the image and iTwin to load the model into application.

We tried using the normal SPA version of the code by generating an App on Bentley platform with type of **SPA**, and when we load iModel to that application using the example which they've given [HERE](https://github.com/iTwin/viewer/tree/master/packages/modules/web-viewer-react) which redirects you to Bentley login page, and after login we are able to show the model. (Used create-react-app command which they mentioned on that page)

But, the application can be used by multiple stakeholder, whom we cant give the Bentley login access to, but they can access and inspect the 3D model which is generated on Bentley platform.

We tried using the repo which you mentioned, [ABOVE](https://github.com/calebmshafer/itwin-viewer-proxy) and tried to use the proxy server to generate the viewer. In order to do that, we had to generate the client ID and Client Secret with the type **Service** in order to use it in `NodeJS` backend to generate token.

Even though we used the code exactly as is in the application, the application is giving the error as `Requested iModel is not available` for this method, and works & loads the model properly on regular login method.

Only difference which we made is, we commented the `changeSetId` since we did not get the exact decoment to generate/get one. Is this mandatory in order to show data in proxy method? if so, please assist us with how to generate the `changeSetId`.

**Reply by aruniverse** [reply]
Did you add `<YOUR_SERVICE_CLIENT_ID>@apps.imsoidc.bentley.com` to your Project?

**Reply by rakesh-agathsya** [reply]
Got it, I did not find this anywhere on the documentation, but had to do some research. We have to invite `service-client-id@apps.imsoidc.bentley.com` as a participant to the iModel created. This solved the issue.

Now, we need to find out how to add participants to the model automatically without the need of manual intervention. If you have the documentation for that, please share. that will be helpful.

**Reply by aruniverse** [reply]
To note, it was included in the readme: 
![image](https://user-images.githubusercontent.com/11051042/179996194-5511141c-cc36-4817-8212-8260f1d03b2e.png)

re, "how to add participants to the model automatically without the need of manual intervention"
You only need to add your service client to the one project, and then as long as all your users are able to obtain a token from the token server (that uses that service client) they should be able to access the imodel

**Reply by rakesh-agathsya** [reply]
Yeah, but it is not clear that we need to add it as participant. anyway, there are multiple iModels. Do I need to add this email to each and every iModel?

**Reply by aruniverse** [reply]
You just need to invite the user to the project where you have the imodels in. If you have multiple projects you need to invite the user to all of them

**Reply by rakesh-agathsya** [reply]
If I login to the Bentley platfoem, I have only 2 options.
1. My Apps &
2. My iModels

Where is project section? Under My iModels, I have all the iModels listed.

**Reply by pmconne** [reply]
It sounds like you are using a sample iModel. The "Create sample iModel" page has this notice:

> Note: Sample iModels should not be used to host sensitive data. The uploaded data is administered by Bentley and should only be used for testing purposes. It is not subject to the same data privacy, security policies, and access controls that apply to Bentley's iTwin offerings. It is subject to data size and other usage limitations, and will be purged after 90 days if not renewed.

Sample iModels belong to a project that is managed by Bentley and are not appropriate for "real" applications. Do you need guidance on how to create a "real" (not sample) iModel that you can manage yourself?
